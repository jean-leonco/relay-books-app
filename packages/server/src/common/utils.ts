import { Model, Types } from 'mongoose';
import { fromGlobalId } from 'graphql-relay';

// https://stackoverflow.com/a/2593661/710693
export const escapeRegex = (str: string) => `${str}`.replace(/[.?*+^$[\]\\(){}|-]/g, '\\$&');

export const PLATFORM = {
  APP: 'APP',
  UNKNOWN: 'UNKNOWN',
};

export const getPlatform = (platform?: string | null): string => {
  if (typeof platform !== 'string') {
    return PLATFORM.UNKNOWN;
  }

  //if (platform.includes('ios') || platform.includes('android')) {
  //  return PLATFORM.APP;
  //}

  if (platform === PLATFORM.APP) {
    return PLATFORM.APP;
  }

  return PLATFORM.UNKNOWN;
};

export function kilometersToEarthRadians(kilometers: number) {
  return kilometers / 6378.1;
}

// returns an ObjectId given an param of unknown type
export const getObjectId = (target: string | Model<any> | Types.ObjectId | undefined | null): Types.ObjectId | null => {
  if (!target) {
    return null;
  }

  if (target instanceof Types.ObjectId) {
    return new Types.ObjectId(target.toString());
  }

  if (typeof target === 'object') {
    return target && target._id ? new Types.ObjectId(target._id) : null;
  }

  if (typeof target === 'string') {
    const result = fromGlobalId(target);

    if (result.type && result.id && Types.ObjectId.isValid(result.id)) {
      return new Types.ObjectId(result.id);
    }

    if (Types.ObjectId.isValid(target)) {
      return new Types.ObjectId(target);
    }

    return null;
  }

  return null;
};

type EnumObject = {
  [key: string]: string;
};

type EnumObjectResult = {
  [key: string]: {
    value: string;
  };
};

export const enumBuilderValues = (constants: EnumObject): EnumObjectResult =>
  Object.keys(constants).reduce(
    (prev, curr) => ({
      ...prev,
      [curr]: {
        value: constants[curr],
      },
    }),
    {},
  );
