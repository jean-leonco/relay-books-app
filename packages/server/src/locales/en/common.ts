const common = {
  SomethingWentWrong: 'Something went wrong.',
  TheDescriptionShouldNot: 'The description must be less than {{size}} characters.',
};
export default common;
export type Keys = keyof typeof common;
