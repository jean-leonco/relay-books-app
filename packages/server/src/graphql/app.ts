/* eslint-disable no-console */
import 'isomorphic-fetch';

import { koaPlayground } from 'graphql-playground-middleware';
import Koa from 'koa';
import bodyParser from 'koa-bodyparser';
import convert from 'koa-convert';
import cors from '@koa/cors';
import graphqlHttp from 'koa-graphql';
import koaLogger from 'koa-logger';
import Router from '@koa/router';
import idx from 'idx';
import i18next from 'i18next';
import koaI18next from 'koa-i18next';
import NoIntrospection from 'graphql-disable-introspection';
import { GraphQLError } from 'graphql';

import * as graphqlLoaders from '../loader';

import { JWT_KEY } from '../common/config';

import { KoaContextExt } from '../types';

import UserModel, { IUser } from '../modules/user/UserModel';

import { t } from '../locales/helpers';

import appPlatformMiddleware from '../modules/auth/appPlatformMiddleware';
import authMiddleware from '../modules/auth/authMiddleware';

import { getDataloaders } from './helper';
import { schema } from './schema';
import { i18nMiddleware } from './i18n';

const app = new Koa<any, KoaContextExt>();

if (process.env.NODE_ENV === 'production') {
  app.proxy = true;
}
app.keys = [JWT_KEY];

const loaders = graphqlLoaders;

const router = new Router<any, KoaContextExt>();

app.use(bodyParser());

app.on('error', (error) => {
  console.error('Error while answering request', { error });
});

if (process.env.NODE_ENV !== 'test') {
  app.use(koaLogger());
}

app.use(convert(cors({ maxAge: 86400, origin: '*' })));

router.get('/healthz', async (ctx) => {
  try {
    await UserModel.find().lean<IUser>();
    ctx.body = 'Database working';
    ctx.status = 200;
  } catch (err) {
    ctx.body = err.toString();
    ctx.status = 500;
  }
});

if (process.env.NODE_ENV !== 'production') {
  router.all(
    '/playground',
    koaPlayground({
      endpoint: '/graphql',
    }),
  );
}

// Middleware to get dataloaders
app.use(async (ctx, next) => {
  ctx.dataloaders = getDataloaders(loaders);
  await next();
});

app.use(
  koaI18next(i18next, {
    lookupCookie: 'lang', // detecting language in cookie
    order: ['header', 'cookie'],
    next: true,
  }),
);

app.use(i18nMiddleware);

// avoid requests from unknown clients
if (process.env.NODE_ENV !== 'development') {
  app.use(appPlatformMiddleware);
}

// Middleware to get result from authorization token
app.use(authMiddleware);

const graphqlSettingsPerReq = async (request, ctx, koaContext) => {
  const { dataloaders, user } = koaContext;
  const { appplatform, timezone } = request.header;

  if (process.env.NODE_ENV !== 'test') {
    console.info('Handling request', {
      appplatform,
      unauthorized: !user,
      userId: idx(user, (_) => _._id),
    });
  }

  return {
    graphiql: process.env.NODE_ENV !== 'production',
    schema,
    //rootValue: { request: ctx.req },
    validationRules: process.env.NODE_ENV === 'production' ? [NoIntrospection] : [],
    context: {
      user,
      dataloaders,
      appplatform,
      koaContext,
      timezone,
      t,
    },
    formatError: (error: GraphQLError) => {
      if (error.name && error.name === 'BadRequestError') {
        ctx.status = 400;
        ctx.body = 'Bad Request';
        return {
          message: 'Bad Request',
        };
      }

      if (error.path || error.name !== 'GraphQLError') {
        console.error(error);
      } else {
        console.log(`GraphQLWrongQuery: ${error.message}`);
      }

      console.error('GraphQL Error', { error });

      if (process.env.NODE_ENV !== 'production') {
        return {
          message: error.message,
          locations: error.locations,
          stack: error.stack,
        };
      } else {
        ctx.status = 400;
        ctx.body = 'Bad Request';
        return {
          message: 'Bad Request',
        };
      }
    },
  };
};

const graphqlServer = graphqlHttp(graphqlSettingsPerReq);
router.all('/graphql', graphqlServer);

app.use(router.routes()).use(router.allowedMethods());

export default app;
