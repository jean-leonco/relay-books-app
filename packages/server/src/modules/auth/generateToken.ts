import jwt from 'jsonwebtoken';
import { addMinutes, addSeconds } from 'date-fns';

import { SESSION_TOKEN_SCOPES, IUser, SessionToken, ISessionToken } from '../../models';
import { GraphQLContext } from '../../types';

import { JWT_KEY } from '../../common/config';

const twelveHoursInMinutes = 60 * 12;

export const getExpirationTimeByScope = (scope: string): number | null => {
  if (Object.values(SESSION_TOKEN_SCOPES).indexOf(scope) === -1) {
    return null;
  }

  switch (scope) {
    case SESSION_TOKEN_SCOPES.SESSION:
      return null;
    case SESSION_TOKEN_SCOPES.RESET_PASSWORD:
      return twelveHoursInMinutes;
    default:
      return null;
  }
};

export const jwtSign = (sessionToken: ISessionToken, secondsToExpire?: number): string => {
  const expiresIn = secondsToExpire || getExpirationTimeByScope(sessionToken.scope);

  return jwt.sign(
    {
      id: sessionToken.user ? sessionToken.user : sessionToken.ip,
      scope: sessionToken.scope,
      sessionToken: sessionToken._id,
    },
    JWT_KEY,
    expiresIn ? { expiresIn: expiresIn * 60 } : {},
  );
};

export const generateToken = async (
  ctx: GraphQLContext,
  user: IUser | null,
  channel: string,
  scope: string,
  expiresIn?: number, // in seconds
): Promise<string | null> => {
  const { koaContext } = ctx;

  const { ip } = koaContext.request;
  const scopeExpirationTime = getExpirationTimeByScope(scope); // in minutes
  const scopeExpiresIn = scopeExpirationTime ? addMinutes(new Date(), scopeExpirationTime) : null;

  let sessionToken;

  // only reuse tokens that has user and the scope is SESSION
  if (user && scope === SESSION_TOKEN_SCOPES.SESSION) {
    sessionToken = await SessionToken.findOne({
      user: user._id,
      ip,
      channel,
      scope,
      isActive: true,
      isBlocked: false,
    }).lean<ISessionToken>();
  }

  if (sessionToken) {
    sessionToken = await SessionToken.findOneAndUpdate(
      { _id: sessionToken._id },
      { expiresIn: expiresIn ? addSeconds(new Date(), expiresIn) : scopeExpiresIn },
    ).lean<ISessionToken>();
  } else {
    sessionToken = await new SessionToken({
      ...(user ? { user: user._id } : {}),
      ip,
      channel,
      scope,
      expiresIn: expiresIn ? addSeconds(new Date(), expiresIn) : scopeExpiresIn,
    }).save();
  }

  return jwtSign(sessionToken!, expiresIn);
};

export const hasReachedMaximumNumberOfSessions = async (user: IUser, scope: string, numberOfSessions: number) => {
  const sessions = await SessionToken.find({
    user: user._id,
    scope,
    isActive: true,
    expiresIn: { $gte: new Date() },
  }).lean();

  return sessions.length >= numberOfSessions;
};
