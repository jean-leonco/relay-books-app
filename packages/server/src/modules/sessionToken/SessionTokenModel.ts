import mongoose, { Document, Model, Types } from 'mongoose';

import { PLATFORM } from '../../common/utils';
import { isActiveMongooseField } from '../../core/mongoose/withMongooseFields';

const { ObjectId } = mongoose.Schema.Types;

export enum SESSION_TOKEN_SCOPES {
  RESET_PASSWORD = 'BOOKSAPP:RESET_PASSWORD', // used when user requested a password reset
  SESSION = 'BOOKSAPP:SESSION', // token for a valid user session
}

const Schema = new mongoose.Schema(
  {
    user: {
      type: ObjectId,
      ref: 'User',
      description: 'User owner of this token.',
    },
    ip: {
      type: String,
      description: 'Agent IP that came from request headers.',
      required: true,
      index: true,
    },
    channel: {
      type: String,
      required: true,
      default: PLATFORM.APP,
      description: 'Channel that this token belongs too. ex: APP',
      enum: Object.values(PLATFORM),
    },
    name: {
      type: String,
      description: 'Name of this token. ex: Jean Linux',
    },
    scope: {
      type: String,
      description: 'Scope of this session token. ex: SESSION',
    },
    expiresIn: {
      type: Date,
      description: 'When this session token expires.',
    },
    isBlocked: {
      type: Boolean,
      description: 'If the token is blocked.',
      default: false,
      required: true,
      index: true,
    },
    ...isActiveMongooseField,
  },
  {
    timestamps: {
      createdAt: 'createdAt',
      updatedAt: 'updatedAt',
    },
    collection: 'SessionToken',
  },
);

export interface ISessionToken extends Document {
  user: Types.ObjectId;
  ip: string;
  channel: string;
  isActive: boolean;
  isBlocked: boolean;
  name: string;
  scope: string;
  expiresIn: Date | null;
  createdAt: Date;
  updatedAt: Date;
}

const SessionTokenModel: Model<ISessionToken> = mongoose.model('SessionToken', Schema);

export default SessionTokenModel;
