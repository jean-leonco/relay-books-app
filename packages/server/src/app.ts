/* eslint-disable no-console */

import cors from '@koa/cors';
import Router from '@koa/router';
import { GraphQLError } from 'graphql';
import { koaPlayground } from 'graphql-playground-middleware';
import i18next from 'i18next';
import Koa from 'koa';
import bodyParser from 'koa-bodyparser';
import graphqlHttp from 'koa-graphql';
import koaI18next from 'koa-i18next';
import logger from 'koa-logger';

import { NODE_ENV, isProduction } from './config';

import { i18nMiddleware } from './i18n';
import { t } from './locales/helpers';
import appPlatformMiddleware from './modules/auth/appPlatformMiddleware';
import authMiddleware from './modules/auth/authMiddleware';
import { getDataloaders } from './modules/loader/loaderRegister';
import schema from './schema/schema';
import { KoaContext } from './types';

const app = new Koa<any, KoaContext>();
const router = new Router<any, KoaContext>();

app.use(bodyParser());
app.use(cors());

app.on('error', (error) => {
  console.error('Error while answering request', { error });
});

if (isProduction) {
  app.proxy = true;
}

if (NODE_ENV !== 'test') {
  app.use(logger());
}

if (!isProduction) {
  router.all('/playground', koaPlayground({ endpoint: '/graphql' }));
}

app.use(async (ctx, next) => {
  ctx.dataloaders = getDataloaders();
  await next();
});

app.use(
  koaI18next(i18next, {
    lookupCookie: 'lang',
    order: ['header', 'cookie'],
    next: true,
  }),
);

app.use(i18nMiddleware);

if (NODE_ENV !== 'development') {
  app.use(appPlatformMiddleware);
}

app.use(authMiddleware);

const graphqlSettingsPerReq = (request, ctx, koaContext) => {
  const { dataloaders, user } = koaContext;
  const { appplatform } = request.header;

  return {
    graphiql: !isProduction,
    schema,
    context: {
      user,
      dataloaders,
      koaContext,
      appplatform,
      t,
    },
    formatError: (error: GraphQLError) => {
      if (error.name && error.name === 'BadRequestError') {
        ctx.status = 400;
        ctx.body = 'Bad Request';
        return {
          message: 'Bad Request',
        };
      }

      if (error.path || error.name !== 'GraphQLError') {
        console.error(error);
      } else {
        console.log(`GraphQLWrongQuery: ${error.message}`);
      }

      console.error('GraphQL Error', { error });

      if (!isProduction) {
        return {
          message: error.message,
          locations: error.locations,
          stack: error.stack,
        };
      } else {
        ctx.status = 400;
        ctx.body = 'Bad Request';
        return {
          message: 'Bad Request',
        };
      }
    },
  };
};

const graphqlServer = graphqlHttp(graphqlSettingsPerReq);
router.all('/graphql', graphqlServer);

app.use(router.routes()).use(router.allowedMethods());

export default app;
