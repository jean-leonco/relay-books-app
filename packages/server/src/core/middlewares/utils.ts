import { GraphQLResolveInfo } from 'graphql';
import { ObjectSchema, Shape } from 'yup';

import { GraphQLContext } from '../../types';

interface GraphQLMutationConfig {
  validationSchema?(context: GraphQLContext): ObjectSchema<Shape<any, any>>;
  authenticatedOnly?: boolean;
}

export function getMutationFields(info: GraphQLResolveInfo, mutationName: string): GraphQLMutationConfig | null {
  const mutationType = info.schema.getMutationType();
  if (!mutationType) {
    return null;
  }

  const fields = mutationType.getFields();
  return fields[mutationName];
}
