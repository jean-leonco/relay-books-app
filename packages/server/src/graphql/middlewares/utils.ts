import { GraphQLField, GraphQLResolveInfo } from 'graphql';

import { Extensions } from '../../types';

interface Field extends GraphQLField<any, any> {
  extensions: Extensions | null | undefined;
}

export function getMutationFields(info: GraphQLResolveInfo, mutationName: string): Field | null {
  const mutationType = info.schema.getMutationType();
  if (!mutationType) {
    return null;
  }

  const fields = mutationType.getFields();
  return fields[mutationName];
}
