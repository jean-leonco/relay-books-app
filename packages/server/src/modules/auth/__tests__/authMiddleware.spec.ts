import { sanitizeTestObject, httpRequestGraphql } from '@booksapp/test-utils';

import {
  clearDbAndRestartCounters,
  connectMongoose,
  createSessionToken,
  createUser,
  disconnectMongoose,
  gql,
} from '../../../../test/helpers';

import app from '../../../graphql/app';
import { jwtSign } from '../generateToken';
import { PLATFORM } from '../../../common/utils';
import SessionTokenModel, { ISessionToken, SESSION_TOKEN_SCOPES } from '../../sessionToken/SessionTokenModel';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

describe('authMiddleware', () => {
  it('should not allow requests without appplatform', async () => {
    const user = await createUser();
    const sessionToken = await createSessionToken({ user });

    const token = jwtSign(sessionToken);

    const query = gql`
      query Q {
        me {
          id
          name
        }
      }
    `;

    const payload = { query, variables: {} };
    const response = await httpRequestGraphql(payload, { authorization: `JWT ${token}` }, app);

    expect(response.status).toBe(403);
    expect(response.body).toMatchSnapshot();
  });

  it('should allow requests with known appplatform', async () => {
    const user = await createUser();
    const sessionToken = await createSessionToken({ user });

    const token = jwtSign(sessionToken);

    const query = gql`
      query Q {
        me {
          id
          name
        }
      }
    `;

    const payload = { query, variables: {} };

    const response = await httpRequestGraphql(
      payload,
      { authorization: `JWT ${token}`, appplatform: PLATFORM.APP },
      app,
    );

    expect(response.status).toBe(200);
    expect(sanitizeTestObject(response.body)).toMatchSnapshot();
  });

  it('should not allow requests with blocked SessionToken', async () => {
    const user = await createUser();
    const sessionToken = await createSessionToken({ user, isBlocked: true });

    const query = gql`
      query Q {
        me {
          email
        }
      }
    `;

    const token = jwtSign(sessionToken);

    const variables = {};
    const payload = {
      query,
      variables,
    };
    const header = {
      authorization: `JWT ${token}`,
      appplatform: PLATFORM.APP,
    };

    const response = await httpRequestGraphql(payload, header, app);

    const sessionTokens = await SessionTokenModel.find().lean<ISessionToken>();
    expect(sessionTokens.length).toBe(1);
    expect(response.status).toBe(401);
    expect(response.body).toMatchSnapshot();
  });

  it('should not allow requests if sessionToken has different scope than SESSION_TOKEN_SCOPES.SESSION', async () => {
    const user = await createUser();
    const sessionToken = await createSessionToken({ user, scope: SESSION_TOKEN_SCOPES.RESET_PASSWORD });

    const query = gql`
      query Q {
        me {
          email
        }
      }
    `;

    const token = jwtSign(sessionToken);

    const variables = {};
    const payload = { query, variables };
    const header = {
      authorization: `JWT ${token}`,
      appplatform: PLATFORM.APP,
    };

    const response = await httpRequestGraphql(payload, header, app);

    const sessionTokens = await SessionTokenModel.find().lean<ISessionToken>();
    expect(sessionTokens.length).toBe(1);
    expect(response.status).toBe(401);
    expect(response.body).toMatchSnapshot();
  });

  it('should allow requests with sessionToken scope being SESSION_TOKEN_SCOPES.SESSION', async () => {
    const user = await createUser();
    const sessionToken = await createSessionToken({ user });

    const query = gql`
      query Q {
        me {
          email
        }
      }
    `;

    const token = jwtSign(sessionToken);

    const variables = {};
    const payload = { query, variables };
    const header = {
      authorization: `JWT ${token}`,
      appplatform: PLATFORM.APP,
    };

    const response = await httpRequestGraphql(payload, header, app);

    const sessionTokens = await SessionTokenModel.find().lean<ISessionToken>();
    expect(sessionTokens.length).toBe(1);
    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot();
  });
});
