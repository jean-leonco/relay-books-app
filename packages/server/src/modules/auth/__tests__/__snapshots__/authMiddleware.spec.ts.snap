// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authMiddleware should allow requests with known appplatform 1`] = `
Object {
  "data": Object {
    "me": Object {
      "id": "FROZEN-ID",
      "name": "User #1",
    },
  },
}
`;

exports[`authMiddleware should allow requests with sessionToken scope being SESSION_TOKEN_SCOPES.SESSION 1`] = `
Object {
  "data": Object {
    "me": Object {
      "email": "user1@example.com",
    },
  },
}
`;

exports[`authMiddleware should not allow requests if sessionToken has different scope than SESSION_TOKEN_SCOPES.SESSION 1`] = `
Object {
  "data": null,
  "errors": Array [
    Object {
      "message": "Invalid token.",
      "severity": "WARNING",
    },
  ],
}
`;

exports[`authMiddleware should not allow requests with blocked SessionToken 1`] = `
Object {
  "data": null,
  "errors": Array [
    Object {
      "message": "Invalid session.",
      "severity": "WARNING",
    },
  ],
}
`;

exports[`authMiddleware should not allow requests without appplatform 1`] = `
Object {
  "data": null,
  "errors": Array [
    Object {
      "message": "Forbidden",
      "severity": "WARNING",
    },
  ],
}
`;
